From: Gerald Schaefer <gerald.schaefer@de.ibm.com>
Subject: s390/pci/dma: use correct segment boundary size
Patch-mainline: v3.14-rc5
Git-commit: 5ec6d4918a45952e99b1b36c93372d79d6927c57
References: bnc#865896, LTC#104566

Description:  s390/pci/dma: use correct segment boundary size
Symptom:      Wrong segment boundary size may result in data corruption.
Problem:      The boundary size for iommu_area_alloc() is currently set to a
              constant value. This is wrong, we shouldn't use a constant value
              but rather the return value of dma_get_seg_boundary(), since a
              device driver can override the default.
Solution:     Use dma_get_seg_boundary().
Reproduction: -

Signed-off-by: Gerald Schaefer <gerald.schaefer@de.ibm.com>
Acked-by: John Jolly <jjolly@suse.de>
---
 arch/s390/pci/pci_dma.c |    8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

--- a/arch/s390/pci/pci_dma.c
+++ b/arch/s390/pci/pci_dma.c
@@ -206,11 +206,13 @@ static void dma_cleanup_tables(struct zp
 	zdev->dma_table = NULL;
 }
 
-static unsigned long __dma_alloc_iommu(struct zpci_dev *zdev, unsigned long start,
-				   int size)
+static unsigned long __dma_alloc_iommu(struct zpci_dev *zdev,
+				       unsigned long start, int size)
 {
-	unsigned long boundary_size = 0x1000000;
+	unsigned long boundary_size;
 
+	boundary_size = ALIGN(dma_get_seg_boundary(&zdev->pdev->dev) + 1,
+			      PAGE_SIZE) >> PAGE_SHIFT;
 	return iommu_area_alloc(zdev->iommu_bitmap, zdev->iommu_pages,
 				start, size, 0, boundary_size, 0);
 }
